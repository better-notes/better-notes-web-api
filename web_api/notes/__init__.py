from . import values
from . import entities
from . import interactors
from . import specs


__all__ = [
    'values',
    'entities',
    'interactors',
    'specs',
]
