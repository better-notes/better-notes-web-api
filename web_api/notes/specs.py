from web_api.commons.specs import Specification


class NoteSpecification(Specification):
    """Return all notes."""
