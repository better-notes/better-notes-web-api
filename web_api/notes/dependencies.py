from fastapi import Depends
from motor import motor_asyncio  # type: ignore
from web_api.notes import interactors, repositories, validators
from web_api.settings import Settings


async def get_mongo_client(
    settings: Settings = Depends(lambda: Settings()),
) -> motor_asyncio.AsyncIOMotorClient:
    # XXX: async b/c motor client requires loop for instantiation ¯\_(ツ)_/¯
    return motor_asyncio.AsyncIOMotorClient(
        settings.MONGO_HOST, settings.MONGO_PORT,
    )


def get_note_repository(
    client: motor_asyncio.AsyncIOMotorClient = Depends(get_mongo_client),
    settings: Settings = Depends(lambda: Settings()),
) -> repositories.NoteRepository:
    return repositories.NoteRepository(client=client, settings=settings)


def get_note_interactor(
    note_repository: repositories.NoteRepository = Depends(
        get_note_repository
    ),
) -> interactors.NoteInteractor:
    return interactors.NoteInteractor(note_repository=note_repository)


def get_paging_validator(
    settings: Settings = Depends(lambda: Settings()),
) -> validators.PagingValidator:
    return validators.PagingValidator(settings=settings)
