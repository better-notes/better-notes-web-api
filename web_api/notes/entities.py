from datetime import datetime

from web_api.accounts.entities import AccountEntity
from web_api.commons import entities
from web_api.notes import values


class NoteEntity(entities.Entity):
    """Aggregate which represents note."""

    id_: str
    account: AccountEntity
    text: str
    tags: list[values.TagValue]
    created_at: datetime
