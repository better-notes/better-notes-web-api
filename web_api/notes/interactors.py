import dataclasses

from typing import List
from web_api.notes import entities, repositories, values, specs


@dataclasses.dataclass
class NoteInteractor:
    note_repository: repositories.NoteRepository = (
        repositories.NoteRepository()
    )

    async def add(
        self, values: List[values.NoteValue]
    ) -> List[entities.NoteEntity]:
        added_value = []
        for value in values:
            added_value.append(await self.note_repository.add(value))
        return added_value

    async def get(self) -> List[entities.NoteEntity]:
        return await self.note_repository.get(specs.ListNoteSpecification())

    async def update(
        self, entities: List[entities.NoteEntity]
    ) -> List[entities.NoteEntity]:
        return await self.note_repository.update(entities)

    async def delete(
        self, entities: List[entities.NoteEntity]
    ) -> List[entities.NoteEntity]:
        return await self.note_repository.delete(entities)
