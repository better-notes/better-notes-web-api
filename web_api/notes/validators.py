import dataclasses

from fastapi import HTTPException, status
from web_api.commons import values
from web_api.settings import Settings


@dataclasses.dataclass
class PagingTooWideException(HTTPException):
    settings: Settings
    status_code = status.HTTP_400_BAD_REQUEST

    @property
    def detail(self) -> str:
        return (
            'Given paging is too wide. '
            f'{self.settings.MAX_PAGING_LIMIT} is allowed.'
        )


@dataclasses.dataclass
class PagingValidator:
    """Validate that given paging doesn't exceed maximum paging limit."""

    settings: Settings

    def validate(self, *, paging: values.Paging) -> None:
        if paging.limit > self.settings.MAX_PAGING_LIMIT:
            raise PagingTooWideException(settings=self.settings)
