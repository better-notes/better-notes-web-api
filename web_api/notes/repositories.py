import dataclasses
from datetime import datetime, timezone

import pymongo
from motor import motor_asyncio
from pymongo.results import InsertOneResult

from web_api.accounts.entities import AccountEntity
from web_api.commons.repositories import AbstractRepository
from web_api.commons.specs import Specification
from web_api.commons.values import OrderingType, Paging
from web_api.notes import entities, values
from web_api.settings import Settings


@dataclasses.dataclass
class NoteRepository(AbstractRepository):
    """Repository for notes."""

    client: motor_asyncio.AsyncIOMotorClient
    settings: Settings

    @property
    def db(self) -> motor_asyncio.AsyncIOMotorDatabase:
        """Mongo database instance."""
        return self.client[self.settings.mongo_database]

    @property
    def notes_collection(self) -> motor_asyncio.AsyncIOMotorCollection:
        """Accounts collection instalce."""
        return self.db[self.settings.notes_collection]

    async def add(
        self, *, account_entity: AccountEntity, note_value_list: list[values.NoteValue],
    ) -> list[entities.NoteEntity]:
        """Add notes into db. Return added notes."""
        inserted_entities = []

        for note_value in note_value_list:
            note_value_dict = note_value.dict()
            created_at = datetime.now(timezone.utc)
            note_insert: InsertOneResult = (
                await self.notes_collection.insert_one(
                    {
                        'account': account_entity.dict(),
                        **note_value_dict,
                        'created_at': created_at,
                    },
                )
            )
            inserted_entities.append(
                entities.NoteEntity(
                    id_=str(note_insert.inserted_id),
                    account=account_entity,
                    **note_value_dict,
                    created_at=created_at,
                ),
            )

        return inserted_entities

    async def get(
        self, *, spec: Specification, paging: Paging, ordering: values.NoteOrdering,
    ) -> list[entities.NoteEntity]:
        """Return notes for given spec."""
        cursor = self.notes_collection.find(spec.get_query()).sort(
            self._get_sorting(ordering=ordering),
        )

        raw_note_list = cursor.limit(paging.limit).skip(paging.offset)
        note_list = []
        async for note in raw_note_list:
            note_list.append(entities.NoteEntity(id_=str(note.pop('_id')), **note))
        return note_list

    async def update(self, *, spec: Specification, note_value: values.NoteValue) -> int:
        """Update notes using spec. Return updated amount."""
        note_value_dict = note_value.dict()

        update_result = await self.notes_collection.update_one(
            spec.get_query(), {'$set': note_value_dict},
        )
        return update_result.modified_count

    async def delete(self, *, spec: Specification) -> int:
        """Delete notes using spec. Return deleted amount."""
        delete_result = await self.notes_collection.delete_many(spec.get_query())
        return delete_result.deleted_count

    def _get_sorting(self, ordering: values.NoteOrdering) -> list[tuple[str, int]]:
        # TODO: move this to ordering value maybe ðŸ¤”.
        sorting = []

        for field, ordering_type in ordering.dict().items():
            if ordering_type == OrderingType.ascending:
                pymongo_ordering_type = pymongo.ASCENDING
            elif ordering_type == OrderingType.descending:
                pymongo_ordering_type = pymongo.DESCENDING
            else:
                raise ValueError('Unknown ordering type {0}'.format(ordering_type))

            sorting.append((field, pymongo_ordering_type))

        return sorting
