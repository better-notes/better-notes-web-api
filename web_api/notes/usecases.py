import dataclasses

import bson

from web_api.accounts.entities import AccountEntity
from web_api.commons.specs import merge_specs
from web_api.commons.values import Paging
from web_api.notes import specs
from web_api.notes.entities import NoteEntity
from web_api.notes.repositories import NoteRepository
from web_api.notes.values import NoteOrdering, NoteValue, TagValue


@dataclasses.dataclass
class NoteInteractor:
    """Interactor for notes."""

    note_repository: NoteRepository

    async def add(
        self, *, account_entity: AccountEntity, note_value_list: list[NoteValue],
    ) -> list[NoteEntity]:
        """Add notes into db. Return added notes."""
        return await self.note_repository.add(
            account_entity=account_entity, note_value_list=note_value_list,
        )

    async def get(
        self,
        *,
        account_entity: AccountEntity,
        paging: Paging,
        ordering: NoteOrdering,
        tag_value_list: list[TagValue],
    ) -> list[NoteEntity]:
        """Return all notes."""
        if tag_value_list:
            spec = merge_specs(
                specs.GetNoteSpecification(username=account_entity.username),
                specs.GetNoteByTagsSpecification(tag_value_list=tag_value_list),
            )
        else:
            spec = specs.GetNoteSpecification(username=account_entity.username)

        return await self.note_repository.get(spec=spec, paging=paging, ordering=ordering)

    async def update(
        self, *, account_entity: AccountEntity, note_entity_list: list[NoteEntity],
    ) -> list[NoteEntity]:
        """Update given notes using id. Return updated notes."""
        for note_entity in note_entity_list:
            await self.note_repository.update(
                spec=specs.UpdateNoteSpecification(
                    username=account_entity.username, _id=bson.ObjectId(note_entity.id_),
                ),
                note_value=NoteValue(**note_entity.dict(exclude={'id_', 'created_at', 'account'})),
            )

        return note_entity_list

    async def delete(
        self, *, account_entity: AccountEntity, note_entity_list: list[NoteEntity],
    ) -> list[NoteEntity]:
        """Delete given notes using id. Return delete notes."""
        object_id_list = []
        for note_entity in note_entity_list:
            object_id_list.append(bson.ObjectId(note_entity.id_))

        await self.note_repository.delete(
            spec=specs.DeleteNoteSpecification(
                username=account_entity.username, object_id_list=object_id_list,
            ),
        )

        return note_entity_list
