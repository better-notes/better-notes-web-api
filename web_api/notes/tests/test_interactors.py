from fastapi.encoders import jsonable_encoder
from syrupy.filters import props

from web_api.notes.tests import factories


class TestNoteInteractor:
    async def test_add_get(self, snapshot) -> None:
        # Given
        note_list = factories.NoteValueFactory.create_batch(3)
        # And
        interactor = factories.NoteInteractorFactory()
        # When
        await interactor.add(note_value_list=note_list)
        note_entity_list = await interactor.get(
            paging=factories.PagingFactory(),
        )
        # Then
        assert jsonable_encoder(note_entity_list) == snapshot(
            exclude=props('id_', 'created_at'),
        )

    async def test_update(self) -> None:
        # Given
        interactor = factories.NoteInteractorFactory()
        note_list = factories.NoteValueFactory.create_batch(3)
        entity_list = await interactor.add(note_value_list=note_list)
        # When
        entity_list[0].text = 'New text'
        await interactor.update(entities=entity_list)
        entity_list = await interactor.get(paging=factories.PagingFactory())
        # Then
        assert entity_list[0].text == 'New text'

    async def test_delete(self) -> None:
        # Given
        interactor = factories.NoteInteractorFactory()
        # And
        note_list = factories.NoteValueFactory.create_batch(3)
        entity_list = await interactor.add(note_value_list=note_list)
        # Ensure:
        assert await interactor.get(paging=factories.PagingFactory()) != []

        # When
        await interactor.delete(entities=entity_list)
        # Then
        assert await interactor.get(paging=factories.PagingFactory()) == []
