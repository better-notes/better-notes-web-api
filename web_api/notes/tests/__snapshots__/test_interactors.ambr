# name: TestNoteInteractor.test_add_get
  <class 'list'> [
    <class 'dict'> {
      'tags': <class 'list'> [
        <class 'dict'> {
          'name': 'tag #0',
        },
      ],
      'text': 'Sample text',
    },
    <class 'dict'> {
      'tags': <class 'list'> [
        <class 'dict'> {
          'name': 'tag #1',
        },
      ],
      'text': 'Sample text',
    },
    <class 'dict'> {
      'tags': <class 'list'> [
        <class 'dict'> {
          'name': 'tag #2',
        },
      ],
      'text': 'Sample text',
    },
  ]
---
